src/core/geometry/qgsgeos.cpp

```diff
@@ -548,7 +587,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
     return distance;
   }
 
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  // GEOSPreparedDistance_r is not able to properly compute the distance if one
+  // of the geometries if a vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))).
+  // In that case, replace `geom` by a single point.
+  // However, GEOSDistance_r works.
+  std::unique_ptr< QgsAbstractGeometry > preparedGeom;
+  if ( mGeosPrepared && isZVerticalLine( geom ) )
```

Comment:
There's an issue here if geom is a multilinestring containing a single line. Best way to handle this is:

```suggestion
  if ( mGeosPrepared && isZVerticalLine( geom->simplifiedTypeRef() ) )
```

Reply:
Done.